internal/pkg/vm/internal/runtime/runtime.go

```diff
@@ -8,27 +8,163 @@ import (
 	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/address"
 )
 
+
+
+// Runtime2 is the VM.
+// Review: the object that implements this outlives actor calls.
+type Runtime2 struct {
+	CurerntEpoch() Epoch
+	// Review: why is this taking an offset? verify size of offset
+	Randomness(epoch Epoch, offset uint64) Randomness
+	// Review: the spec still has a TODO on this
+	// Review: not sure why the original has a messagereceipt instead of an InvocOutput
+	Send(input InvocInput) InvocOutput
+	// Review: how does this work? What qualifies as an "error"?
+	// Review: how is this any different than the caller just ignoring a returned error?
+	// SendAllowingErrors(input msg.InvocInput) msg.MessageReceipt
+
+	// Review: we might want a global Storage() Storage here
+}
+
+// InvocationContext is passed to the actors on each method call.
+// Review: this type + Runtime2 is mostly what the spec defines as Runtime.
+type InvocationContext interface {
+	// Review: grouped the VM level things out of the invocation itself
+	Runtime() Runtime2
+	Caller() address.Address
+	// Review: (is in spec, dont think we need it) idem to Caller().Equal(addr) :/
+	// ValidateCallerIs(caller addr.Address)
+	// Review: not sure what all this pattern matching adds up to, the "singleton" pattern does an equality check
+	// ValidateCallerMatches(CallerPattern)
+	// Review: the spec handles state and state changes a bit differently
+	// Review: our Storage is a bit of a mix of AcquireState and the Ipld{Get-Put}
+	// AcquireState() ActorStateHandle
+
+	// Review: The spec suggest handling return construction by the runtime, interesting but not strictly necessary
+	// SuccessReturn()     msg.InvocOutput
+	// ValueReturn(Bytes)  msg.InvocOutput
+	// ErrorReturn(exitCode exitcode.ExitCode) msg.InvocOutput
+
+	// Review: how does this work? is the spec expecting us to Panic? or just terminate the message execution?
+	// Review: this should be `Abort(string) -> msg.InvocOutput` if we follow the pattern
+	// Abort(string)
+
+	// Review: this is a bit trickier than abort in Go. This is a conditional early exit on the function which we cannot do.
+	// Review: `if <cond> { return AssertFailure("blah") }` is a bit closer to go if we are to try to follow this pattern.
+	// Assert(bool)
+
+	// Review: we didnt have this before, why does the actor need to know the value supplied?
+	ValueSupplied() types.AttoFIL
+
+	// Review: if this is just for the `init` actor then it should not be here.
+	// Review: the init actor is a special system actor, its ok for it to have a different interface with a bit more power.
```

Comment:
Yes, the intent is that `CreateActor` is only callable by the init actor (enforced by the runtime). I believe it simplifies the implementation if we can treat the init actor like any other actor in this regard, but if it turns out to be easier to make it a special case, I don't see any particular problems this would cause (@zixuanzh @jbenet anything I've missed here?)

Reply:
Sounds good, we will probably keep it separate on our implementation so other actors don't see it.